uqcsbot/dominos_coupons.py

```diff
@@ -0,0 +1,153 @@
+from typing import Optional
+import logging
+import requests
+from requests.exceptions import RequestException
+from bs4 import BeautifulSoup
+
+import discord
+from discord import app_commands
+from discord.ext import commands
+
+from uqcsbot.bot import UQCSBot
+
+
+MAX_COUPONS = 10  # Prevents abuse
+COUPONESE_DOMINOS_URL = "https://www.couponese.com/store/dominos.com.au/"
+
+
+class DominosCoupons(commands.Cog):
+    def __init__(self, bot: UQCSBot):
+        self.bot = bot
+
+    @app_commands.command()
+    @app_commands.describe(
+        number_of_coupons="The number of coupons to return. Defaults to 5 with max 10.",
+        ignore_expiry="Indicates to include coupons that have expired. Defaults to True.",
+        keywords="Words to search for within the coupon. All coupons descriptions will mention at least one keyword.",
+    )
+    async def dominoscoupons(
+        self,
+        interaction: discord.Interaction,
+        number_of_coupons: Optional[int] = 5,
+        ignore_expiry: Optional[bool] = True,
+        keywords: Optional[str] = "",
+    ):
+        """
+        Returns a list of dominos coupons
+        """
+        await interaction.response.defer(thinking=True)
+
+        if number_of_coupons < 1 or number_of_coupons > MAX_COUPONS:
+            await interaction.edit_original_response(
+                content=f"You can't have that many coupons. Try a number between 1 and {MAX_COUPONS}."
+            )
+            return
+        try:
+            coupons = _get_coupons(number_of_coupons, ignore_expiry, keywords.split())
+        except RequestException as error:
+            logging.warning(
+                f"Could not connect to dominos coupon site ({COUPONESE_DOMINOS_URL}): {error.response.content}"
+            )
+            await interaction.edit_original_response(
+                content=f"Sadly could not reach the coupon website (<{COUPONESE_DOMINOS_URL}>)..."
+            )
+            return
+        except HTTPResponseException as error:
+            logging.warning(
+                f"Received a HTTP response code that was not OK (200), namely ({error.http_code}). Error information: {error.message}"
+            )
+            await interaction.edit_original_response(
+                content=f"Could not find the coupons on the coupon website (<{COUPONESE_DOMINOS_URL}>)..."
+            )
+            return
+
+        if not coupons:
+            await interaction.edit_original_response(
+                content=f"Could not find any coupons matching the given arguments from the coupon website (<{COUPONESE_DOMINOS_URL}>)."
+            )
+            return
+
+        message = "Domino's Coupons:\n"
+        for coupon in coupons:
+            message += f"`{coupon.code}` - {coupon.description} *[Expires: {coupon.expiry_date}]*\n"
+        await interaction.edit_original_response(content=message)
+
+
+class Coupon:
+    def __init__(self, code: str, expiry_date: str, description: str) -> None:
```

Comment:
I question the utility of adding a coupon class when you're just manually accessing the fields anyway, but I guess there's not really a better way without getting into horrific kwarg-generator bullshit for `embed.add_field`.

I guess you could have a method that chains embeds (i.e. something like `embed = Coupon.add_self_to_embed(embed)`) but that's up to you.

Reply:
I'm going to skip on it. The class is definitely overkill, but it seems better than passing a dictionary around. Also, it is only put into an embed once, so I'll leave creating that method to someone else.